### PR TITLE
discoverd: Register

### PR DESCRIPTION
**Pull request by archSeer**
Incremental patch towards getting rid of sdutil.

/cc @titanous

### CODE REVIEW

**Comment by archSeer** [comment]
All statements are covered except `Addr()`, which I added to satisfy the heartbeater interface.

File: discoverd/health/register_test.go [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+			steps: []step{
+				{event: true, up: true, register: true, success: false},
+				{register: true, success: false},
+				{event: true},
+				{}, // make sure register does not run
+				{},
+			},
+		},
+	} {
+		fmt.Println("--- TEST:", t.name, "---")
```
Comment (titanous) [inline discoverd/health/register_test.go]: Use `c.Log`

File: discoverd/health/register_test.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+			steps: []step{
+				{setMeta: true, success: false},
+			},
+		},
+		{
+			name: "setmeta while erroring registration",
+			steps: []step{
+				{event: true, up: true, register: true, success: false},
+				{register: true, success: false},
+				{setMeta: true, success: false},
```
Comment (titanous) [inline discoverd/health/register_test.go]: This and the above check should ensure that when register succeeds the new meta is reflected.

File: discoverd/health/register_test.go [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+type RegistrarFunc func(service string, inst *discoverd.Instance) (discoverd.Heartbeater, error)
+
+func (f RegistrarFunc) RegisterInstance(service string, inst *discoverd.Instance) (discoverd.Heartbeater, error) {
+	return f(service, inst)
+}
+
+type HeartbeatFunc func(map[string]string) error
+
+func (f HeartbeatFunc) SetMeta(meta map[string]string) error { return f(meta) }
+func (f HeartbeatFunc) Close() error                         { return nil }
```
Comment (titanous) [inline discoverd/health/register_test.go]: Do we have any tests that the registration is closed when the service goes down?

**Comment by archSeer** [comment]
@titanous Everything addressed, also added a test for `Addr()`, so we now have 100% coverage.

File: discoverd/health/register_test.go (hunk 2 above)
Comment (titanous) [inline discoverd/health/register_test.go]: Was this addressed?

Comment (archSeer) [inline discoverd/health/register_test.go]: [Yes.](https://github.com/flynn/flynn/pull/913/files#diff-2e1f378428df428e27bf8eabe9dc247eR167)

Comment (titanous) [inline discoverd/health/register_test.go]: It doesn't look like the necessary register step is called after the failing setmeta?

Comment (archSeer) [inline discoverd/health/register_test.go]: Ah true, nice catch. Fixed.

File: discoverd/health/register_test.go [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+			steps: []step{
+				{event: true, up: true, register: true, success: true},
+				{setMeta: true, success: true},
+			},
+		},
+		{
+			name: "setmeta while offline",
+			steps: []step{
+				{setMeta: true, success: false},
+			},
```
Comment (titanous) [inline discoverd/health/register_test.go]: It think this also needs an extra step or two. `setMeta` while offline should just save locally, it should not call the underlying `SetMeta` func

File: discoverd/health/register_test.go [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+				// Addr should be nil now
+				c.Assert(hb.Addr(), Equals, "")
+			} else {
+				wait++
+				errCheck(unregisterChan, stop)
+			}
+			if step.setMeta {
+				go func() {
+					call := <-metaChan
+					call.returnVal <- step.success
```
Comment (titanous) [inline discoverd/health/register_test.go]: Does this ever happen if the heartbeater is "offline" (unregistered)?

File: discoverd/health/register_test.go (hunk 4 above)
Comment (archSeer) [inline discoverd/health/register_test.go]: Fixed as well, we now call register afterwards to confirm that the correct meta is transmitted.

File: discoverd/health/register_test.go (hunk 5 above)
Comment (archSeer) [inline discoverd/health/register_test.go]: Oh, good point... The logic for this is getting fairly complicated though, I think it should be obvious from the implementation that SetMeta will work correctly.

**Comment by titanous** [comment]
Okay, LGTM. Make sure you've rebased on master before merging.

**Pull request merged**